### PR TITLE
Feature/unit tests

### PR DESCRIPTION
**Pull request by shahadat3669**
### Project requirements
- Add unit tests for Book class
- Add unit tests for Person class
- Add unit tests for Student class
- Add unit tests for Teacher class
- Add unit tests for Classroom class
- Add unit tests for Rental class

### General requirements
- There are [no linter errors](https://github.com/microverseinc/linters-config).
- Used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Documented work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
- Follow list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

### CODE REVIEW

File: src/spec/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    expect(@book).to be_instance_of(Book)
+  end
+
+  it 'it has a title of First Book' do
+    expect(@book.title).to eq(title)
+  end
+
+  it 'it has a author of First Author' do
+    expect(@book.author).to eq(author)
+  end
```
Comment (tresorsawasawa) [inline src/spec/book_spec.rb:21]: - [ ] As required for the project to have unit tests for all the `methods` in all the `classes`, Kindly ensure that you add tests on the `rented_by` method in the `Book` class

File: src/spec/classroom_spec.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  it 'It should return correct number of student in a classroom' do
+    stud1 = Student.new('student 1', 20, parent_permission: true, classroom: @classroom)
+    @classroom.add_student(stud1)
+    stud2 = Student.new('student 2', 22, parent_permission: false, classroom: @classroom)
+    @classroom.add_student(stud2)
+    stud3 = Student.new('student 3', 20, parent_permission: true, classroom: @classroom)
+    @classroom.add_student(stud3)
+
+    expect(@classroom.students.length).to eq(3)
+  end
```
Comment (tresorsawasawa) [inline src/spec/classroom_spec.rb:28]: - [ ] As required for the project to have unit tests for all the `methods` in all the `classes`, Kindly ensure that you add specific tests on the `add_student` method in the `Classroom` class. I can see you use the `add_student` method in the above tests for other purpose **not for testing the method itself(`add_student` method)**

File: src/spec/person_spec.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    expect(@person.age).to eq(age)
+  end
+
+  it 'Person has parent permission' do
+    expect(@person.parent_permission).to eq(parent_permission)
+  end
+
+  it 'Person correct_name method should return correctly' do
+    expect(@person.correct_name).to eql(name)
+  end
```
Comment (tresorsawasawa) [inline src/spec/person_spec.rb:31]: - [ ] As required for the project to have unit tests for all the `methods` in all the `classes`, Kindly ensure that you add tests to the `#can_use_services?`, `#of_age?`, and `#rent` methods in the `Person` class as well.

File: src/spec/student_spec.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    expect(@student.age).to eq(age)
+  end
+
+  it 'Student has parent permission' do
+    expect(@student.parent_permission).to eq(parent_permission)
+  end
+
+  it 'Student has a class named First Class' do
+    expect(@student.classroom).to eql(classroom)
+  end
```
Comment (tresorsawasawa) [inline src/spec/student_spec.rb:31]: - [ ] As required for the project to have unit tests for all the `methods` in all the `classes`, Kindly ensure that you add tests to the `#add_classroom?`, and `#play_hooky` methods in the `Student` class as well.

File: src/spec/teacher_spec.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    expect(@teacher.age).to eq(age)
+  end
+
+  it 'Teacher has parent permission' do
+    expect(@teacher.parent_permission).to eq(parent_permission)
+  end
+
+  it 'Teacher should have specialization named First specialization' do
+    expect(@teacher.specialization).to eql(specialization)
+  end
```
Comment (tresorsawasawa) [inline src/spec/teacher_spec.rb:31]: - [ ] As required for the project to have unit tests for all the `methods` in all the `classes`, Kindly ensure that you add tests on the `can_use_services?` method in the `Teacher` class as well

File: src/spec/capitalize_spec.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+require_relative '../capitalize_decorator'
+require_relative '../person'
+
+describe CapitalizeDecorator do
+  it 'it should return name in caplitalize form' do
+    person = Person.new(30, 'shahadat hossain')
+    capitalize_person = CapitalizeDecorator.new(person)
+    expect(capitalize_person.correct_name).to eql('Shahadat hossain')
+  end
+end
```
Comment (tresorsawasawa) [inline src/spec/capitalize_spec.rb:10]: - [ ] As required for the project, Kindly make sure you add tests to all classes that you created while implementing the **Decorator** design pattern. Right now, I noticed that some classes miss tests with their `methods` respectively, I suggest you add them(i.e. tests for `Decorator`, `Nameable`, classes and `their methods respectively` )

Comment (tresorsawasawa) [inline src/spec/capitalize_spec.rb:10]: - [ ] As required for the project, Kindly make sure you add tests to all classes that you created while implementing the **Decorator** design pattern. Right now, I noticed that some classes miss tests with their `methods` respectively, I suggest you add them(i.e. tests for `Decorator`, `Nameable` classes and `their methods respectively` )

- [ ] `README.md` file: Kindly make sure you add all your project's contributors in the `Authors` section of your README and a command on how `to run your tests` in the appropriate section. For short remind just refer the **[README Template](https://github.com/microverseinc/readme-template)**

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @GutemaG and @shahadat3669  
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

almost-there

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
- All tests are passing :heavy_check_mark: 
- Follow Gitflow :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by GutemaG** [comment]
@tresorsawasawa thank you very much for your detailed explanations.

**Review by DeliceLydia — APPROVED** [review]
## Hi Guys, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
# Congratulations! 🎉 

## Highlights
- Addressed all changes requested


### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**